internal/kafka/internal/config/kafka_supported_instance_types.go

```diff
@@ -0,0 +1,202 @@
+
+		if err := kafkaInstanceSize.validate(kp.Id); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type KafkaInstanceSize struct {
+	Id                          string   `yaml:"id"`
+	IngressThroughputPerSec     Quantity `yaml:"ingressThroughputPerSec"`
+	EgressThroughputPerSec      Quantity `yaml:"egressThroughputPerSec"`
+	TotalMaxConnections         int      `yaml:"totalMaxConnections"`
+	MaxDataRetentionSize        Quantity `yaml:"maxDataRetentionSize"`
+	MaxPartitions               int      `yaml:"maxPartitions"`
+	MaxDataRetentionPeriod      string   `yaml:"maxDataRetentionPeriod"`
+	MaxConnectionAttemptsPerSec int      `yaml:"maxConnectionAttemptsPerSec"`
+	QuotaConsumed               int      `yaml:"quotaConsumed"`
+	QuotaType                   string   `yaml:"quotaType"`
+	CapacityConsumed            int      `yaml:"capacityConsumed"`
+}
+
+// validates Kafka instance size configuration to ensure the following:
+//
+// - all properties must be defined
+// - any non-id string values must be parseable
+// - any int values must not be less than or equal to zero
+func (k *KafkaInstanceSize) validate(instanceTypeId string) error {
+	if k.EgressThroughputPerSec.IsEmpty() || k.IngressThroughputPerSec.IsEmpty() ||
+		k.MaxDataRetentionPeriod == "" || k.MaxDataRetentionSize.IsEmpty() || k.Id == "" || k.QuotaType == "" {
+		return fmt.Errorf("Kafka instance size '%s' for instance type '%s' is missing required parameters.", k.Id, instanceTypeId)
+	}
+
+	egressThroughputQuantity, err := k.EgressThroughputPerSec.ToK8Quantity()
+	if err != nil {
+		return fmt.Errorf("egressThroughputPerSec for Kafka instance type '%s', size '%s' is invalid: %s", k.Id, instanceTypeId, err.Error())
+	}
+
+	ingressThroughputQuantity, err := k.IngressThroughputPerSec.ToK8Quantity()
+	if err != nil {
+		return fmt.Errorf("ingressThroughputPerSec for Kafka instance type '%s', size '%s' is invalid: %s", k.Id, instanceTypeId, err.Error())
+	}
+
+	maxDataRetentionSize, err := k.MaxDataRetentionSize.ToK8Quantity()
+	if err != nil {
+		return fmt.Errorf("maxDataRetentionSize for Kafka instance type '%s', size '%s' is invalid: %s", k.Id, instanceTypeId, err.Error())
+	}
+
+	maxDataRetentionPeriod, err := duration.ParseISO8601(k.MaxDataRetentionPeriod)
+	if err != nil {
+		return fmt.Errorf("maxDataRetentionPeriod for Kafka instance type '%s', size '%s' is invalid: %s", k.Id, instanceTypeId, err.Error())
+	}
+
+	if maxDataRetentionPeriod.IsZero() || egressThroughputQuantity.CmpInt64(1) < 0 ||
+		ingressThroughputQuantity.CmpInt64(1) < 0 || maxDataRetentionSize.CmpInt64(1) < 0 ||
+		k.TotalMaxConnections <= 0 || k.MaxPartitions <= 0 || k.MaxConnectionAttemptsPerSec <= 0 ||
+		k.QuotaConsumed < 1 || k.CapacityConsumed < 1 {
+		return fmt.Errorf("Kafka instance size '%s' for instance type '%s' specifies a property value less than or equals to Zero.", k.Id, instanceTypeId)
+	}
+
+	return nil
+}
+
+type SupportedKafkaInstanceTypesConfig struct {
```

Comment:
struct definitions are better to be placed at the start of the file along with other definitions

Reply:
I think this is a personal preference matter.

For example I tend to find useful to have the struct type definition and then all its methods following it instead of placing the struct definitions at the start of the file.

In the Go source code itself you can observe examples on how they follow the "struct type definition and then all its methods" approach instead of placing all struct definitions at the top. See for example https://go.dev/src/image/geom.go

Reply:
Agreed its a personal preference, but i saw in KFM in other places that was followed, thought that was a practice :)

Reply:
Not that I am aware. Marking this thread as resolved as it is a design preference